documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc

```diff
@@ -1113,17 +1115,124 @@ public interface OrderMapper {
 
     OrderMapper INSTANCE = Mappers.getMapper( OrderMapper.class );
 
-    @Mappings({
-        @Mapping(source = "EXTRA", target = "SPECIAL"),
-        @Mapping(source = "STANDARD", target = "DEFAULT"),
-        @Mapping(source = "NORMAL", target = "DEFAULT")
+    @ValueMappings({
+        @ValueMapping(source = "EXTRA", target = "SPECIAL"),
+        @ValueMapping(source = "STANDARD", target = "DEFAULT"),
+        @ValueMapping(source = "NORMAL", target = "DEFAULT")
+    })
+    ExternalOrderType orderTypeToExternalOrderType(OrderType orderType);
+}
+----
+====
+
+.Enum mapping method result
+====
+[source, java, linenums]
+[subs="verbatim,attributes"]
+----
+// GENERATED CODE
+public class OrderMapperImpl implements OrderMapper {
+
+    @Override
+    public ExternalOrderType orderTypeToExternalOrderType(OrderType orderType) {
+        if ( orderType == null ) {
+            return null;
+        }
+
+        ExternalOrderType externalOrderType_;
+
+        switch ( orderType ) {
+            case EXTRA: externalOrderType_ = ExternalOrderType.SPECIAL;
+            break;
+            case STANDARD: externalOrderType_ = ExternalOrderType.DEFAULT;
+            break;
+            case NORMAL: externalOrderType_ = ExternalOrderType.DEFAULT;
+            break;
+            case RETAIL: externalOrderType_ = ExternalOrderType.RETAIL;
+            break;
+            case B2B: externalOrderType_ = ExternalOrderType.B2B;
+            break;
+            default: throw new IllegalArgumentException( "Unexpected enum constant: " + orderType );
+        }
+
+        return externalOrderType_;
+    }
+}
+----
+====
+By default an error will be raised by MapStruct in case a constant of the source enum type does not have a corresponding constant with the same name in the target type and also is not mapped to another constant via `@ValueMapping`. This ensures that all constants are mapped in a safe and predictable manner. An IllegalStateException is thrown when for some reason an unrecognized source value occurs.
```

Comment:
I don't think this was changed in this PR, but I would expect the generated code to throw an Exception if the source enum was expanded without regenerating the code. I don't think there is something else can we do there? :-/